### ISSUE TITLE
Bug: key-auth plugin validating OPTIONS requests

### ISSUE THREAD

**Issue by kiwicopple** [Mashape/kong#852]
The Key-Auth plugin is expecting the api-key header on OPTIONS requests, however this header is stripped out by browsers (Chrome) before request is made.

Details about CORS & OPTIONS requests are explained (much better than I can) here: http://stackoverflow.com/a/15734032

**Comment by thefosk** [comment]
@kiwicopple Can you try to also install the CORS plugin? https://getkong.org/plugins/cors/

**Comment by kiwicopple** [comment]
@thefosk apologies for the glacial response time.

For reference, the original error (without CORS plugin) was this:
**Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8100' is therefore not allowed access. The response had HTTP status code 401.**

I can enable the CORS plugin with the following settings:
```
{
      "enabled": true,
      "name": "cors",
      "config": {
        "methods": [
          "GET",
          "POST"
        ],
        "credentials": true,
        "max_age": 3600,
        "preflight_continue": false,
        "headers": [
          "origin",
          "content-type",
          "accept",
          "Authorization",
          "Access-Control-Allow-Headers",
          "Access-Control-Allow-Origin"
        ]
      }
    }
```
Which will give the following error:
**XMLHttpRequest cannot load <url>. The 'Access-Control-Allow-Origin' header contains multiple values '*, ' is therefore not allowed access.**

Which I believe is related to #259 . It seems that the plugin is attaching a new header rather than concatenating to the existing one?

**Comment by kiwicopple** [comment]
Just an update, I got around the **XMLHttpRequest cannot load . The 'Access-Control-Allow-Origin' header contains multiple values '*, ' is therefore not allowed access** issue, but I had to make some modifications to the plugin CORS plugin itself. I won't submit a pull request because I don't think it is the correct logic, but below are the steps I took to correct it on my server at runtime -

Enable CORS:
```
curl -X POST $host/apis/$api/plugins \
   --data "name=cors" \
   --data "config.methods=GET, POST" \
   --data "config.credentials=true" \
   --data "config.max_age=3600" \
   --data "config.preflight_continue=true" \
   --data "config.headers=origin, content-type, accept, Authorization, Access-Control-Allow-Headers, Access-Control-Allow-Origin, X-Total-Count" \
   --data "config.exposed_headers=origin, content-type, accept, Authorization, Access-Control-Allow-Headers, Access-Control-Allow-Origin, X-Total-Count"
```

Notice that I **don't** have **--data "config.origin=*"** set.

I then edited this file on my server:
/usr/local/share/lua/5.1/kong/plugins/cors/access.lua

and I changed this:
```
local function configure_origin(ngx, conf)
  if conf.origin == nil then
    ngx.header["Access-Control-Allow-Origin"] = "*"
  else
    ngx.header["Access-Control-Allow-Origin"] = conf.origin
    ngx.header["Vary"] = "Origin"
  end
end
```

to this:
```
local function configure_origin(ngx, conf)
  if conf.origin ~= nil then
    ngx.header["Access-Control-Allow-Origin"] = conf.origin
    ngx.header["Vary"] = "Origin"
  end
end
```
This means that it won't add a new header unless the conf.origin is explicitly set when enabling the plugin, and any headers sent from the server are just passed through.

Just a bit of background - my app is running SailsJs, and is on AWS Beanstalk (so it sits behind an nginx server). Between the two of them, I couldn't figure out how to NOT send the Access-Control-Allow-Origin headers. Since the Kong CORS plugin didn't seem to want to concatenate, I just choose to remove the  Access-Control-Allow-Origin header completely if "config.origin" is not set. I think the actual fix should be to concatenate whatever is sent from the server (or to override it completely), however I will leave that to the Pros to decide

**Comment by thefosk** [comment]
This issue may be closed by #1240

**Comment by thefosk** [comment]
@kiwicopple can you try with Kong v.0.8.2+ ?
